### DISCUSSION TITLE
Should Agent could expose "clients" property?

### DISCUSSION THREAD

**Discussion by StefanCenusa** [nodejs/undici#1996] [Q&A] [answered]
Hello,

Recently I've tried switching from NodeJS keep-alive agent to undici and one thing that I was unable to do without hacking undici was to access the clients created by an Agent instance. ([this property](https://github.com/nodejs/undici/blob/main/lib/agent.js#L57))

My Agent talks to many origins and I need to see the PoolStats on each of those Pools that it creates.

My current solution was to replicate the code of the Agent (the initialisation and handling of clients), something like this:
```
    #request(rawRequest) {
        const key = String(rawRequest.origin);
        const ref = this.#pools.get(key);

        let pool = ref ? ref.deref() : null;
        if (!pool) {
            pool = new Pool(rawRequest.origin, this.#poolOptions);
            this.#pools.set(key, new WeakRef(pool));
            this.#finalizer.register(pool, key);
        }

        const poolStats = pool.stats;
        if (poolStats.queued > 0) {
            throw new Error('queue full');
        }

        return pool.request(rawRequest);
    }
```
however, if I had acces to the clients map (or client stats) I could have avoided this hacky approach.

I think my topic relates a bit to [this one](https://github.com/nodejs/undici/discussions/1254).

What do you think it's the right approach here?

Thanks!

**Comment by ronag** [comment]
What do you need the pool stats for? Wouldn't diagnostic channels be a solution?

**Comment by StefanCenusa** [comment]
I don't know what "diagnostic channels" are, but I need the PoolStats (or stats) for the clients created by the agent. There can be multiple reasons for this requirement. 

Monitoring can be one (exposing stats for each origin to a prometheus endpoint), but in my current use-case I need those stats to make a client-side circuit breaking:
-  if I see that an origin is saturated (has max in-flight requests), don't send it more requests, drop the request; as previously shown in the code snippet:
```
        const poolStats = pool.stats;
        if (poolStats.queued > 0) {
            throw new Error('queue full');
        }
 ```

**Reply by ronag** [reply]
If it's full then `dispatch` will return false and then later emit a drain event.

**Reply by StefanCenusa** [reply]
Ok, but is it recommended to use Agent.dispatch since it is a low level API?

**Reply by StefanCenusa** [reply]
Also, it seems that when using `dispatch` I should define a `DispatchHandler` which looks like an overhead in terms of usability.

**Reply by ronag** [reply]
You don't need to do anything, all the other methods should work as they call dispatch.

**Comment by ronag** [comment, marked as answer]
```js
class MyPool extends Pool {
  #full = false
  constructor () {
    super()
    this.on('drain', () => { this.#full = false })
  }
   dispatch(...args) {
     const ret = super.dispatch(...args)
     this.#full = !ret
     return ret
   }
}
const dispatcher = new MyPool()
undici.request(url, { dispatcher })
```

**Reply by StefanCenusa** [reply]
Oh, really cool. I'll give it a shot. Many thanks for the pointers!
